src/models/chan.js

```diff
@@ -183,21 +179,85 @@ Chan.prototype.getFilteredClone = function(lastActiveChannel, lastMessage) {
 	}, {});
 };
 
-function writeUserLog(client, msg) {
-	if (!msg.isLoggable()) {
-		return false;
+Chan.prototype.writeUserLog = function(client, msg) {
+	this.messages.push(msg);
+
+	// Does this user have logs disabled
+	if (!client.config.log) {
+		return;
+	}
+
+	// Are logs disabled server-wide
+	if (Helper.config.messageStorage === "none") {
+		return;
 	}
 
+	// Is this particular message or channel loggable
+	if (!msg.isLoggable() || !this.isLoggable()) {
+		return;
+	}
+
+	// Find the parent network where this channel is in
 	const target = client.find(this.id);
 
 	if (!target) {
-		return false;
+		return;
 	}
 
-	userLog.write(
-		client.name,
-		target.network.host, // TODO: Fix #1392, multiple connections to same server results in duplicate logs
-		this.type === Chan.Type.LOBBY ? target.network.host : this.name,
-		msg
-	);
-}
+	// TODO: Something more pluggable
+	if (Helper.config.messageStorage === "sqlite") {
+		client.messageStorage.index(target.network.uuid, this.name, msg);
+	} else if (Helper.config.messageStorage === "text") {
+		userLog.write(
+			client.name,
+			target.network.host, // TODO: Fix #1392, multiple connections to same server results in duplicate logs
+			this.type === Chan.Type.LOBBY ? target.network.host : this.name,
+			msg
+		);
+	}
+};
+
+Chan.prototype.loadMessages = function(client, network) {
+	if (!this.isLoggable()) {
+		return;
+	}
+
+	client.messageStorage
+		.getMessages(network, this)
+		.then((messages) => {
+			if (messages.length === 0) {
+				if (network.irc.network.cap.isEnabled("znc.in/playback")) {
+					this.requestZncPlayback(network, 0);
+				}
+
+				return;
+			}
+
+			// TODO: This needs to insert correctly based on time
+			this.messages.unshift(...messages);
+
+			if (!this.firstUnread) {
+				this.firstUnread = messages[messages.length - 1].id;
+			}
+
+			client.emit("more", {
+				chan: this.id,
+				messages: messages.slice(-100),
+			});
+
+			if (network.irc.network.cap.isEnabled("znc.in/playback")) {
+				const from = Math.floor(messages[messages.length - 1].time.getTime() / 1000);
```

Comment:
So is this from the last message we have in the logs until present?

Reply:
Yep.